polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                continue 'values;
+            }
+        }
+        // Create new variable if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        eq_values.insert(new_id, value);
+    }
+
+    let mut contained_values = HashMap::new();
+    'contained_values: for (value, var) in var_info.contained_values {
+        for (id, set) in &mut variables {
+            if set.contains(&var) {
+                contained_values
+                    .entry(id.clone())
+                    .or_insert_with(HashSet::new)
+                    .insert(value);
+                continue 'contained_values;
+            }
+        }
+        // Create new variable if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        let mut new_val_set = HashSet::new();
+        new_val_set.insert(value);
+        contained_values.insert(new_id, new_val_set);
+    }
+
+    let mut types = HashMap::new();
+    'types: for (var, typ) in var_info.types {
+        for (id, set) in &mut variables {
+            if set.contains(&var) {
+                // @TODO(steve): If we already have a type for it make sure they match don't just
+                // overwrite it.
+                types.insert(id.clone(), typ);
+                continue 'types;
+            }
+        }
+        // Create new variable if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        types.insert(new_id, typ);
+    }
+
+    let mut this_id = String::new();
+    for (id, set) in &variables {
+        if set.contains(&Symbol::new("_this")) {
+            this_id = id.clone()
+        }
+    }
+
+    Vars {
+        variables,
+        field_relationships,
+        in_relationships,
+        eq_values,
+        contained_values,
+        types,
+        this_id,
+    }
+}
+
+fn constrain_vars(types: &Types, vars: &Vars, this_type: &str) -> ResultSet {
+    let mut result_set = ResultSet {
+        requests: HashMap::new(),
+        resolve_order: vec![],
+        result_id: vars.this_id.clone(),
+    };
+    constrain_var(&mut result_set, types, vars, &vars.this_id, this_type);
+    result_set
+}
+
+fn constrain_var(
+    result_set: &mut ResultSet,
+    types: &Types,
+    vars: &Vars,
+    var_id: &str,
+    var_type: &str,
+) {
+    // @TODO(steve): Probably should check the type against the var types. I think???
+    let type_def = types
+        .iter()
+        .find(|r| r.0 == var_type)
+        .map(|r| r.1.clone())
+        .unwrap_or_else(HashMap::new);
+
+    let mut request = if result_set.requests.contains_key(var_id) {
+        result_set.requests.remove(var_id).unwrap()
```

Comment:
wouldn't this branch result in pushing the `var_id` to the resolve order twice?

Reply:
Hmm yeah, gotta fix resolve order in general, It's not right. Will make that a followup task.